tests/src/app/maptooladdfeatureline/testqgsmaptooladdfeatureline.cpp

```diff
@@ -851,5 +865,75 @@ void TestQgsMapToolAddFeatureLine::testStreamTolerance()
 }
 
+void TestQgsMapToolAddFeatureLine::testDifferentCrs()
+{
+
+  TestQgsMapToolAdvancedDigitizingUtils utils( mCaptureTool );
+  QSet<QgsFeatureId> oldFids = utils.existingFeatureIds();
+  mCanvas->setDestinationCrs( QgsCoordinateReferenceSystem( QStringLiteral( "EPSG:3946" ) ) );
+
+  // Prepare geometries
+  QgsPolylineXY line1;
+  line1 << QgsPointXY( 0, 0 ) << QgsPointXY( 0, 10 );
+  QgsFeature lineF1;
+  lineF1.setGeometry( QgsGeometry::fromPolylineXY( line1 ) );
+
+  mLayerCRS3946Line->startEditing();
+  mLayerCRS3946Line->addFeature( lineF1 );
+  QgsFeatureId mFidLine3946 = lineF1.id();
+  QCOMPARE( mLayerCRS3946Line->featureCount(), ( long )1 );
+  // just one added feature
+  QCOMPARE( mLayerCRS3946Line->undoStack()->index(), 1 );
+
+  // Draw the line to avoid some error with the projections
+  mLayerCRS3945Line->startEditing();
+  mCanvas->setCurrentLayer( mLayerCRS3945Line );
+  utils.mouseClick( 8, 2, Qt::LeftButton, Qt::KeyboardModifiers(), true );
+  utils.mouseClick( 8, 8, Qt::LeftButton, Qt::KeyboardModifiers(), true );
+  utils.mouseClick( 8, 8, Qt::RightButton );
+  QgsFeatureId mFidLine3945 = utils.newFeatureId( oldFids );
+  QCOMPARE( mLayerCRS3945Line->featureCount(), ( long )1 );
+  // just one added feature
+  QCOMPARE( mLayerCRS3945Line->undoStack()->index(), 1 );
+
+
+  mCanvas->setCurrentLayer( mLayerCRS3946Line );
+
+  QgsSnappingConfig cfg = mCanvas->snappingUtils()->config();
+  cfg.setEnabled( true );
+  cfg.setMode( QgsSnappingConfig::AllLayers );
+  cfg.setTypeFlag( QgsSnappingConfig::SegmentFlag );
+  cfg.setTolerance( 50 );
+  cfg.setUnits( QgsTolerance::Pixels );
+
+  bool topologicalEditing = cfg.project()->topologicalEditing();
+  cfg.project()->setTopologicalEditing( true );
+
+  mCanvas->snappingUtils()->setConfig( cfg );
+
+  utils.mouseClick( -2, 4, Qt::LeftButton, Qt::KeyboardModifiers(), true );
+  utils.mouseClick( 0, 4, Qt::LeftButton, Qt::KeyboardModifiers(), true );
+  utils.mouseClick( 2, 4, Qt::LeftButton, Qt::KeyboardModifiers(), true );
+  utils.mouseClick( 2, 4, Qt::RightButton );
+
+  QString new3946Wkt = "LineString (0 0, 0 4, 0 10)";
+
+  QCOMPARE( mLayerCRS3946Line->getFeature( mFidLine3946 ).geometry().asWkt( 0 ), new3946Wkt ); // use wkt to avoid float rounding errors
+  mLayerCRS3946Line->undoStack()->undo();
+
+  // same test with a layer with a different CRS
+  mCanvas->setCurrentLayer( mLayerCRS3945Line );
+  utils.mouseClick( 0, 4, Qt::LeftButton, Qt::KeyboardModifiers(), true );
+  utils.mouseClick( 8, 4, Qt::LeftButton, Qt::KeyboardModifiers(), true );
+  utils.mouseClick( 8, 4, Qt::RightButton );
+
+  // (2 2, 8 2, 8 4, 8 8) transformed from 3946 to 3945
+  QString new3945Wkt = "LineString (17441.5859966475982219 -862119.21677098423242569, 17441.58100318093784153 -862117.23846332356333733, 17441.57101624645292759 -862113.28184797242283821)";
+  QCOMPARE( mLayerCRS3945Line->getFeature( mFidLine3945 ).geometry().asWkt( 2 ), QgsGeometry::fromWkt( new3945Wkt ).asWkt( 2 ) );
```

Comment:
Judging by the coordinates in the comment (2 2, 8 2, 8 4, 8 8) (there is no such linestring in the test), something probably got mixed up here. This doesn't test the PR's proposed functionality. You should test that topological points were added to the 3946 layer when digitizing on the 3945.
The test as it is is already passing in master without the PR!